neqo-transport/src/cc/cubic.rs

```diff
@@ -0,0 +1,205 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// CUBIC congestion control
+#![deny(clippy::pedantic)]
+#![allow(clippy::cast_precision_loss)]
+#![allow(clippy::cast_possible_truncation)]
+#![allow(clippy::cast_sign_loss)]
+
+use std::cmp::{max, min};
+use std::fmt::{self, Display};
+use std::time::{Duration, Instant};
+
+use crate::cc::{CWND_MIN, MAX_DATAGRAM_SIZE_F64};
+use neqo_common::qtrace;
+
+const CUBIC_C: f64 = 0.4;
+const CUBIC_BETA: f64 = 0.7;
+const CUBIC_FAST_CONERGENCE: f64 = 0.85; // (1 + CUBIC_BETA)/ 2
+pub const CUBIC_BETA_USIZE: usize = 70;
+pub const CUBIC_DIV: usize = 100;
+const CUBIC_ALFA: f64 = 3.0 * (1.0 - 0.7) / (1.0 + 0.7);
+
+#[derive(Debug)]
+pub struct Cubic {
+    last_max_cwnd: f64,
+    estimated_tcp_cwnd: f64,
+    k: f64,
+    w_max: f64,
+    ca_epoch_start: Option<Instant>,
+    last_phase_was_tcp: bool,
+}
+
+impl Default for Cubic {
+    fn default() -> Self {
+        Self {
+            last_max_cwnd: 0.0,
+            estimated_tcp_cwnd: 0.0,
+            k: 0.0,
+            w_max: 0.0,
+            ca_epoch_start: None,
+            last_phase_was_tcp: false,
+        }
+    }
+}
+
+impl Display for Cubic {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Cubic [last_max_cwnd: {}, k: {}, w_max: {}, ca_epoch_start: {:?}]",
+            self.last_max_cwnd, self.k, self.w_max, self.ca_epoch_start
+        )?;
+        Ok(())
+    }
+}
+
+impl Cubic {
+    // W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1)
+    fn w_cubic(&self, t: f64) -> f64 {
+        CUBIC_C * (t - self.k).powi(3) * MAX_DATAGRAM_SIZE_F64 + self.w_max
+    }
+
+    // K = cubic_root(W_max*(1-beta_cubic)/C) (Eq. 2)
+    fn calc_k(&mut self) {
+        self.k = libm::cbrt(self.w_max / MAX_DATAGRAM_SIZE_F64 * (1.0 - CUBIC_BETA) / CUBIC_C);
+    }
+
+    pub fn calculate_cwnd_ca(
+        &mut self,
+        congestion_window: &mut usize,
+        now: Instant,
+        rtt: Duration,
+        acked_bytes: &mut usize,
+        newly_acked_bytes: usize,
+    ) {
+        let curr_cwnd = *congestion_window as f64;
+        self.last_phase_was_tcp = false;
+        let curr_cwnd_f64 = *congestion_window as f64;
+        if self.ca_epoch_start.is_none() {
+            // This is the first time the congestion avoidance has been entered.
```

Comment:
"first time" implies that this happens only once, but this is just the start of the epoch, which happens every time we exit recovery.